manta-crypto/src/signature/mod.rs

```diff
@@ -191,50 +209,68 @@ where
 pub mod schnorr {
     use super::*;
     use crate::{
-        algebra::{security::DiscreteLogarithmHardness, Group, Scalar},
-        eclair::{bool::Bool, cmp::PartialEq, Has},
+        algebra::{
+            security::DiscreteLogarithmHardness, Group as _, HasGenerator, Ring, ScalarMul,
+            ScalarMulGroup,
+        },
+        eclair::{
+            alloc::{Const, Constant},
+            bool::Bool,
+            cmp::PartialEq,
+            Has,
+        },
         hash::security::PreimageResistance,
+        rand::{Rand, RngCore, Sample},
     };
     use core::{cmp, fmt::Debug, hash::Hash, marker::PhantomData};
 
     /// Schnorr Signature Hash Function
-    pub trait HashFunction<G, COM = ()>: PreimageResistance
-    where
-        G: DiscreteLogarithmHardness + Group<COM>,
-    {
+    pub trait HashFunction<COM = ()>: PreimageResistance {
+        /// Scalar Type
+        type Scalar: Ring<COM>;
+
+        /// Group Type
+        type Group: ScalarMulGroup<Self::Scalar, COM> + DiscreteLogarithmHardness;
+
         /// Message Type
         type Message;
 
-        /// Hashes `message` along with `verifying_key` and `nonce_point` into a
-        /// [`Scalar`](Group::Scalar).
+        /// Hashes `message` along with `verifying_key` and `nonce_point` into a scalar of type
+        /// [`Scalar`](Self::Scalar.
         fn hash(
             &self,
-            verifying_key: &G,
-            nonce_point: &G,
+            verifying_key: &Self::Group,
+            nonce_point: &Self::Group,
             message: &Self::Message,
             compiler: &mut COM,
-        ) -> G::Scalar;
+        ) -> Self::Scalar;
     }
 
+    /// Scalar Type
+    pub type Scalar<H, COM = ()> = <H as HashFunction<COM>>::Scalar;
+
+    /// Group Type
+    pub type Group<H, COM = ()> = <H as HashFunction<COM>>::Group;
+
+    /// Message Type
+    pub type Message<H, COM = ()> = <H as HashFunction<COM>>::Message;
+
     /// Schnorr Signature
     #[derive(derivative::Derivative)]
     #[derivative(
-        Clone(bound = "G::Scalar: Clone, G: Clone"),
-        Copy(bound = "G::Scalar: Copy, G: Copy"),
-        Debug(bound = "G::Scalar: Debug, G: Debug"),
-        Eq(bound = "G::Scalar: Eq, G: Eq"),
-        Hash(bound = "G::Scalar: Hash, G: Hash"),
-        PartialEq(bound = "G::Scalar: cmp::PartialEq, G: cmp::PartialEq")
+        Clone(bound = "S: Clone, G: Clone"),
+        Copy(bound = "S: Copy, G: Copy"),
+        Debug(bound = "S: Debug, G: Debug"),
+        Eq(bound = "S: Eq, G: Eq"),
+        Hash(bound = "S: Hash, G: Hash"),
+        PartialEq(bound = "S: cmp::PartialEq, G: cmp::PartialEq")
     )]
-    pub struct Signature<G, COM = ()>
-    where
-        G: DiscreteLogarithmHardness + Group<COM>,
-    {
+    pub struct Signature<S, G> {
```

Comment:
This is the same as the other point about adding the `COM` argument. We want to avoid adding it to this structure which would be needed. It's not `G: ScalarMulGroup<S>` that applies here but `G: ScalarMulGroup<S, COM>`.